src/api/artifact/abstractor.go

```diff
@@ -0,0 +1,112 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package artifact
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/distribution/manifest/manifestlist"
+	"github.com/docker/distribution/manifest/schema1"
+	"github.com/docker/distribution/manifest/schema2"
+	"github.com/goharbor/harbor/src/api/artifact/blob"
+	"github.com/goharbor/harbor/src/api/artifact/resolver"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+	"github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// Abstractor abstracts the specific information for different types of artifacts
+type Abstractor interface {
+	// Abstract the specific information for the specific artifact type into the artifact model,
+	// the information can be got from the manifest or other layers referenced by the manifest.
+	// The "repoFullName" is the "project_name/repository_name" that the artifact belongs to
+	Abstract(ctx context.Context, repoFullName string, artifact *artifact.Artifact) error
+}
+
+// NewAbstractor returns an instance of the default abstractor
+func NewAbstractor() Abstractor {
+	return &abstractor{
+		blobReader: blob.NewReader(),
+	}
+}
+
+type abstractor struct {
+	blobReader blob.Reader
+}
+
+// TODO try CNAB, how to forbid CNAB
+
+// TODO add white list for supported artifact type
+func (a *abstractor) Abstract(ctx context.Context, repoFullName string, artifact *artifact.Artifact) error {
+	// read manifest content
+	manifestMediaType, content, err := a.blobReader.Read(repoFullName, artifact.Digest, true)
+	if err != nil {
+		return err
+	}
+	artifact.ManifestMediaType = manifestMediaType
+
+	switch artifact.ManifestMediaType {
+	// docker manifest v1
```

Comment:
What if we do not support it?

Reply:
If we don't support it, the images with v1 manifest cannot be pushed into Harbor, and there may be some issue when upgrading if the images with v1 manifest already exist in Harbor 1.x.

Reply:
this is good question, I noticed that docker has deprecated the schemav1 support.